source/lib.ts

```diff
@@ -0,0 +1,312 @@
+	const middleware = handleAsyncErrors(
+		async (
+			request: Express.Request,
+			response: Express.Response,
+			next: Express.NextFunction,
+		) => {
+			// First check if we should skip the request
+			const skip = await Promise.resolve(options.skip(request, response))
+			if (skip) {
+				next()
+				return
+			}
+
+			// Create an augmented request
+			const augmentedRequest = request as AugmentedRequest
+
+			// Get a unique key for the client
+			const key = await Promise.resolve(options.keyGenerator(request, response))
+			// Increment the client's hit counter by one
+			const { totalHits, resetTime } = await options.store.increment(key)
+
+			// Get the quota (max number of hits) for each client
+			const retrieveQuota =
+				typeof options.max === 'function'
+					? options.max(request, response)
+					: options.max
+
+			const maxHits = await Promise.resolve(retrieveQuota)
+			// Set the rate limit information on the augmented request object
+			augmentedRequest[options.requestPropertyName] = {
+				limit: maxHits,
+				current: totalHits,
+				remaining: Math.max(maxHits - totalHits, 0),
+				resetTime,
+			}
+
+			// Set the X-RateLimit headers on the response object if enabled
+			if (options.legacyHeaders && !response.headersSent) {
+				response.setHeader('X-RateLimit-Limit', maxHits)
+				response.setHeader(
+					'X-RateLimit-Remaining',
+					augmentedRequest[options.requestPropertyName].remaining,
+				)
+
+				// If we have a resetTime, also provide the current date to help avoid issues with incorrect clocks
+				if (resetTime instanceof Date) {
+					response.setHeader('Date', new Date().toUTCString())
+					response.setHeader(
+						'X-RateLimit-Reset',
+						Math.ceil(resetTime.getTime() / 1000),
+					)
+				}
+			}
+
+			// Set the standardized RateLimit headers on the response object
+			// if enabled
+			if (options.standardHeaders && !response.headersSent) {
+				response.setHeader('RateLimit-Limit', maxHits)
+				response.setHeader(
+					'RateLimit-Remaining',
+					augmentedRequest[options.requestPropertyName].remaining,
+				)
+
+				if (resetTime) {
+					const deltaSeconds = Math.ceil(
+						(resetTime.getTime() - Date.now()) / 1000,
+					)
+					response.setHeader('RateLimit-Reset', Math.max(0, deltaSeconds))
+				}
+			}
+
+			// If we are to skip failed/successfull requests, decrement the
+			// counter accordingly once we know the status code of the request
+			if (options.skipFailedRequests || options.skipSuccessfulRequests) {
+				let decremented = false
+				const decrementKey = async () => {
+					if (!decremented) {
+						await options.store.decrement(key)
+						decremented = true
+					}
+				}
+
+				if (options.skipFailedRequests) {
+					response.on('finish', async () => {
+						if (!options.requestWasSuccessful(request, response))
+							await decrementKey()
+					})
+					response.on('close', async () => {
+						if (!response.writableEnded) await decrementKey()
```

Comment:
I believe this was previously 

`if (!res.finished) {`

Why did it change to `writeEnabled`?

Reply:
It was deprecated in Node 13.4.0 - https://nodejs.org/api/http.html#http_request_finished

Reply:
Aah, good catch. Thanks!